src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java

```diff
@@ -118,6 +116,89 @@ public synchronized void close() throws IOException {
+                // Must detach from current console before attaching to another
+                if (!k32.FreeConsole()) {
+                    logger.atError(STOP_GRACEFULLY_EVENT).log("FreeConsole error {}", k32.GetLastError());
+                }
+                // Attach to the console that's running the target process
+                if (!k32.AttachConsole(pid)) {
+                    logger.atError(STOP_GRACEFULLY_EVENT).log("AttachConsole error {}", k32.GetLastError());
+                    return;
+                }
+                // Make gg ignore Ctrl-C temporarily so that we don't terminate ourselves as well
+                if (!Kernel32Ex.INSTANCE.SetConsoleCtrlHandler(null, true)) {
+                    logger.atError(STOP_GRACEFULLY_EVENT)
+                            .log("SetConsoleCtrlHandler add error {}", k32.GetLastError());
+                    return;
+                }
+                // Send Ctrl-C to all processes in the console
+                if (k32.GenerateConsoleCtrlEvent(Wincon.CTRL_C_EVENT, 0)) {
+                    sentConsoleCtrlEvent = true;
+                    // Wait to ensure CtrlHandler is not enabled before the calling process gets the signal
+                    try {
+                        Thread.sleep(2000);
+                    } catch (InterruptedException ignore) {
+                        logger.atWarn(STOP_GRACEFULLY_EVENT).log("CtrlHandler sleep interrupted");
+                        Thread.currentThread().interrupt();
+                    }
+                } else {
+                    logger.atError(STOP_GRACEFULLY_EVENT)
+                            .log("GenerateConsoleCtrlEvent error {}", k32.GetLastError());
+                }
+            } finally {
+                // Re-attach gg to original console and re-enable Ctrl-C
+                if (!k32.FreeConsole()) {
+                    logger.atError(STOP_GRACEFULLY_EVENT).log("FreeConsole error {}", k32.GetLastError());
+                }
+                int holderPid = Processes.newPidProcess(holderProc).getPid();
+                if (!k32.AttachConsole(holderPid)) {
+                    logger.atError(STOP_GRACEFULLY_EVENT).log("Re-AttachConsole error {}", k32.GetLastError());
+                }
+                if (!Kernel32Ex.INSTANCE.SetConsoleCtrlHandler(null, false)) {
+                    logger.atError(STOP_GRACEFULLY_EVENT)
+                            .log("Re-SetConsoleCtrlHandler error {}", k32.GetLastError());
+                }
+                holderProc.destroyForcibly();
+            }
+        }
+
+        if (sentConsoleCtrlEvent) {
+            try {
+                long timeToWaitInSec = Math.max(gracefulShutdownTimeout.getSeconds() - 2, 0);
+                process.waitFor(timeToWaitInSec, TimeUnit.SECONDS);
+                logger.debug("Process stopped gracefully: {}", pid);
+            } catch (InterruptedException ignore) {
+                Thread.currentThread().interrupt();
+            }
```

Comment:
instead of catching this I think I would throw it up and just let the `finally` do the force kill

Reply:
We will have to catch InterruptedException in close() method anyway. Is there any difference between catching it here vs catching it in close() ?

Reply:
I'm actually only worried about what I told you to add in re-interrupting the thread. That may mean that taskkill won't be started since I don't know that Java would start a process from an interrupted thread. But we really need to be sure that taskkill kills the process. We could also use `TerminateProcess` JNA call instead of taskkill since we own the process anyway.

Reply:
We still rely on taskkill to terminate the whole process tree. Let's just throw it up?

Reply:
Yep